### PR TITLE
WIP: Add dask-like config object

### PR DESCRIPTION
**Pull request by djhoese**
See https://github.com/dask/dask/pull/4099 for more information. This PR is the start of a solution to allow satpy users to configure things interactively using the same interface that dask uses. Things like this would be possible:

```python
with satpy.config.set(chunk_size=2048):
    scn = Scene(reader='X', filenames=[...])
    scn.load(['C01', 'C02'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets()
```

Chunk size is the first obvious option to make available to users since it is already used as a global configuration value. Others could be `ppp_config_dir` (with optionally better name), default resampling algorithm, etc.

 - [ ] Tests added 
 - [ ] Tests passed 
 - [ ] Passes ``git diff origin/master -- "*py" | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
Ah the tests are going to fail because they are pytest tests and we are using unittest.

**Comment by mraspaud** [comment]
While the `with` syntax is useful in many cases, I would like to see also the following available:
```python
satpy.config.set(chunk_size=2048)
scn = Scene(reader='X', filenames=[...])
scn.load(['C01', 'C02'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets()
```
to set the option globally.

Also, I was wondering about saving/loading the configuration from file, eg to a `.satpyrc`. Is something like this planned ?

**Comment by djhoese** [comment]
Both of those features already exist. Details to come...

**Comment by djhoese** [comment]
@mraspaud See this code: https://github.com/pytroll/satpy/pull/461/files#diff-9638ada601a9fda1b6e7cf3e4ce56befR280

These are used in the `collect_yaml` function which will search for the configuration files in order. The `collect_env` function is then used to look for `SATPY_*` environment variables. These are then "merged" in order by the `merge` function the later items in the list of yaml/env configs where the `env` come last and have the most priority.

Note: there are also some fancy environment variable handlings involved where things like `__` (double underscore) can be used to specify a sub-dictionary of configuration (ex. `SATPY_FOO__BAR_BAZ=5` would set `config_obj.config['foo']['bar-baz'] = 5`.

File: satpy/_config.py [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+    {'x': 2, 'y': {'a': 2, 'b': 3}}
+
+    >>> a = {'x': 1, 'y': {'a': 2}}
+    >>> b = {'x': 2, 'y': {'b': 3}}
+    >>> update(a, b, priority='old')  # doctest: +SKIP
+    {'x': 1, 'y': {'a': 2, 'b': 3}}
+
+    See Also
+    --------
+    dask.config.merge
```
Comment (mraspaud) [inline satpy/_config.py]: This needs to be changes to `satpy.config.merge`, right ?

File: satpy/_config.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    Examples
+    --------
+    >>> a = {'x': 1, 'y': {'a': 2}}
+    >>> b = {'y': {'b': 3}}
+    >>> merge(a, b)  # doctest: +SKIP
+    {'x': 1, 'y': {'a': 2, 'b': 3}}
+
+    See Also
+    --------
+    dask.config.update
```
Comment (mraspaud) [inline satpy/_config.py]: same as above

File: satpy/_config.py [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+            data = yaml.load(f.read()) or {}
+            configs.append(data)
+
+    return configs
+
+
+def collect_env(prefix, env=None):
+    """Collect config from environment variables
+
+    This grabs environment variables of the form "DASK_FOO__BAR_BAZ=123" and
```
Comment (mraspaud) [inline satpy/_config.py]: same here

File: satpy/_config.py [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+
+    Examples
+    --------
+    >>> import dask
+    >>> with dask.config.set({'foo': 123}):
+    ...     pass
+
+    See Also
+    --------
+    dask.config.get
```
Comment (mraspaud) [inline satpy/_config.py]: here too

File: satpy/_config.py [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+        env : dict
+            The system environment variables
+
+        Returns
+        -------
+        config: dict
+
+        See Also
+        --------
+        dask.config.refresh: collect configuration and update into primary config
```
Comment (mraspaud) [inline satpy/_config.py]: and here

File: satpy/_config.py [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+
+        >>> config.get('foo.x')  # doctest: +SKIP
+        1
+
+        >>> config.get('foo.x.y', default=123)  # doctest: +SKIP
+        123
+
+        See Also
+        --------
+        dask.config.set
```
Comment (mraspaud) [inline satpy/_config.py]: here...

**Comment by mraspaud** [comment]
Also I though you wanted this as a standalone package ?

**Comment by djhoese** [comment]
@mraspaud Read the linked to dask PR. Yes this will become another package. This PR is a WIP and a proof of concept. The documentation has not been updated yet. This PR will likely be closed and a new one opened using the new package if/when it exists. First, I wanted to make a PR to dask to see what stuff of theirs would break. I figured adoption would be quicker that way.

**Comment by mraspaud** [comment]
sounds good !

**Comment by djhoese** [comment]
I'm closing this because the new `donfig` package has been created which will supersede any changes already made in this PR. I'll make a new PR when I roll donfig in.

**Pull request closed without merging**